File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.StringTokenizer;

/*
 * BOJ_1697
 * - 수빈이는 동생과 숨바꼭질
 * - 수빈이는 걷거나 순간이동 가능
 * - 만약 수빈이의 위치가 X 일 때 걷는 다면, 1초 후에 x-1. x+ 1 이동
 * - 순간이동을 하는 경우 2 * x
 * 수빈이와 동생 위치가 주어질 때 수빈이가 동생을 찾는 가장 빠른 방법?
 * 
 * - N과 K가 0, 10만 과 같은 상황이라면, DFS일 경우 쓸데없이 많은 연산처리가 되는 것이죠,,, 실제로 제출해보면 시간초과가 뜹니다.
 * 
 * - 그래서 BFS로 풀었습니다. 
 * 
 * */

public class Main {
	
	static BufferedReader br;
	static StringTokenizer st;

	static int N, K;
	static int result;
	
	static int [] distance; // 거리 전체를 놓고 ! (메모이제이션 기법)

	public static void bfs(int start) {
		
		Deque<Integer> queue = new ArrayDeque<Integer>();
		queue.add(start); // 큐에 시작 위치 넣기
		distance[start] = 0;
		
		while (!queue.isEmpty()) {
			int currentIdx = queue.poll();
			
			// 만약 동생을 찾으면,
			if (currentIdx == K) {
				System.out.println(distance[currentIdx]);
				return;
			}

			for (int nextIdx : new int[] {currentIdx+1, currentIdx-1, currentIdx * 2}) {	
				if (0 > nextIdx || nextIdx >= 100001) {
					continue;
				}
					
				// 처음 방문했을경우
				if (distance[nextIdx] == 0) {
					distance[nextIdx] = distance[currentIdx] + 1;
					queue.add(nextIdx);
				}
				
			}
			
		}
		
	}

	public static void main(String[] args) throws IOException{
		// TODO Auto-generated method stub

		br = new BufferedReader(new InputStreamReader(System.in));
		st = new StringTokenizer(br.readLine().trim());
		
		N = Integer.parseInt(st.nextToken());
		K = Integer.parseInt(st.nextToken());
		
		distance = new int[100001];
		bfs(N);
	}

}

File: 백준/Silver/1697. 숨바꼭질/숨바꼭질.py
"""
수빈 N & 동생 J
수빈 : 걷거나 순간이동 가능
- 걷기 : 수빈위치가 X일 때 걸으면 1초 후 X - 1, X + 1로 이동
- 순간이동 : 1초 후 2*X로 이동
결과 : 동생을 찾을 수 있는 가장 빠른 시간 몇 초 후?
-> 전략: 백트래킹
"""
from collections import deque

N, K = map(int, input().split())
MAX = 100000
distance = [0] * (MAX + 1)

def bfs(start):
    queue = deque([start])
    while queue:
        x = queue.popleft() # current index
        if x == K:
            return distance[x]
        for nx in (x+1, x-1, x*2):
            if 0 <= nx <= MAX and not distance[nx]:
                distance[nx] = distance[x] + 1
                queue.append(nx)

result = bfs(N)
print(result)